types/react-router/index.d.ts

```diff
@@ -104,11 +104,9 @@ export type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P i
 export type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
 
 export function matchPath<P>(pathname: string, props: RouteProps): match<P> | null;
-export function withRouter<P extends RouteComponentProps<any>>(component: React.ComponentType<P>): React.ComponentClass<Omit<P, keyof RouteComponentProps<any>>>;
 
-// decorator signature
 // There is a known issue in TypeScript, which doesn't allow decorators to change the signature of the classes
 // they are decorating. Due to this, if you are using @withRouter decorator in your code,
 // you will see a bunch of errors from TypeScript. The current workaround is to use withRouter() as a function call
 // on a separate line instead of as a decorator.
-export function withRouter<P, TFunction extends React.ComponentClass<P>>(target: TFunction): TFunction;
+export function withRouter<P extends RouteComponentProps<any>>(component: React.ComponentType<P>): React.ComponentClass<Omit<P, keyof RouteComponentProps<any>>>;
```

Comment:
@Grmiade to resolve confusion, could you please add a test case that demonstrates intended usage and that was failing before your changes?

Reply:
I would rather change that to:
```typescript
export function withRouter<P>(component: React.ComponentType<P & RouteComponentProps<any>>): React.ComponentClass<P>;
```
And remove unnencessary `Omit` in this definition.

Reply:
The `Omit` definition allows to infer the type of the resulting component. We don't have to remove it. It has been introduce with the PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21329 to fix https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17181.

Reply:
The first `withRouter` definition is good. The problem is the second definition. Look the [comment](https://github.com/Grmiade/DefinitelyTyped/blob/83a1193dc47592c2ac86b4d7119cf70ffe135129/types/react-router/index.d.ts#L108-L111). The second definition is useless :/